python/ray/serve/tests/test_batching.py

```diff
@@ -58,6 +60,265 @@ def __call__(self, requests):
+
+    class FloatIntegerBatch:
+        @serve.batch(max_batch_size=1.0)
+        async def method(self, requests):
+            pass
+
+    with pytest.raises(ValueError):
+
+        class NegativeTimeout:
+            @serve.batch(batch_wait_timeout_s=-0.1)
+            async def method(self, requests):
+                pass
+
+    class FloatZeroTimeout:
+        @serve.batch(batch_wait_timeout_s=0.0)
+        async def method(self, requests):
+            pass
+
+    class IntZeroTimeout:
+        @serve.batch(batch_wait_timeout_s=0)
+        async def method(self, requests):
+            pass
+
+    with pytest.raises(TypeError):
+
+        class NonTimeout:
+            @serve.batch(batch_wait_timeout_s="a")
+            async def method(self, requests):
+                pass
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("use_class", [True, False])
+async def test_batch_size_one_long_timeout(use_class):
+    @serve.batch(max_batch_size=1, batch_wait_timeout_s=1000)
+    async def long_timeout(requests):
+        if "raise" in requests:
+            1 / 0
+        return requests
+
+    class LongTimeout:
+        @serve.batch(max_batch_size=1, batch_wait_timeout_s=1000)
+        async def long_timeout(self, requests):
+            if "raise" in requests:
+                1 / 0
+            return requests
+
+    cls = LongTimeout()
+
+    async def call(arg):
+        if use_class:
+            return await cls.long_timeout(arg)
+        else:
+            return await long_timeout(arg)
+
+    assert await call("hi") == "hi"
+    with pytest.raises(ZeroDivisionError):
+        await call("raise")
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("use_class", [True, False])
+async def test_batch_size_multiple_zero_timeout(use_class):
+    @serve.batch(max_batch_size=2, batch_wait_timeout_s=0)
+    async def zero_timeout(requests):
+        await asyncio.sleep(1)
+        if "raise" in requests:
+            1 / 0
+        return requests
+
+    class ZeroTimeout:
+        @serve.batch(max_batch_size=2, batch_wait_timeout_s=0)
+        async def zero_timeout(self, requests):
+            await asyncio.sleep(1)
+            if "raise" in requests:
+                1 / 0
+            return requests
+
+    cls = ZeroTimeout()
+
+    async def call(arg):
+        if use_class:
+            return await cls.zero_timeout(arg)
+        else:
+            return await zero_timeout(arg)
+
+    assert await call("hi") == "hi"
+    with pytest.raises(ZeroDivisionError):
+        await call("raise")
+
+    # Check that 2 requests will be executed together if available.
+    # The first should cause a size-one batch to be executed, then
+    # the next two should be executed together (signaled by both
+    # having the exception).
+    t1 = asyncio.get_event_loop().create_task(call("hi1"))
+    await asyncio.sleep(0.5)
+    t2 = asyncio.get_event_loop().create_task(call("hi2"))
+    t3 = asyncio.get_event_loop().create_task(call("raise"))
```

Comment:
I'm a bit confused here, why are these executed in a size-two batch if timeout = 0?  Doesn't timeout=0 mean that the max batch size is effectively 1?

Reply:
The second two are executed together because they are both waiting once the first one finishes (only one batch is executed at a time)

Reply:
Oh right, thanks

Reply:
Oh right, thanks